### ISSUE TITLE
Visual="Material" breaks certain controls

### ISSUE THREAD

**Issue by lewcianci** [xamarin/Xamarin.Forms#5692]
### Description

When using Visual="Material", certain controls don't work anymore.

### Steps to Reproduce

1. Install CarouselView from AlexRainman
2. Put CarouselView in a Frame on a Page that has Visual="Material" up the top
3. You will get a NullReferenceException on the ItemChanged event

### Expected Behavior

CarouselView renders fine

### Actual Behavior

App crashes due to NullReferenceException

### Basic Information

- Version with issue: 3.6.0.264807
- Last known good version: N/A
- IDE: Visual studio for Mac
- Platform Target Frameworks: 
  - iOS:  11.1 (Works on iOS)
  - Android:  9.0
  - UWP:  N/A
- Android Support Library Version: 
- Nuget Packages:
- Affected Devices:

#### This is an issue with AlexRainman's CarouselView, why didn't you open an issue on his GitHub page?
In my mind, something as ubiquitous as a "Frame" should always work, no matter what "Visual" properties are applied to it.  Changing the "Visual" attribute of a Frame should not ever cause a control inside of the Frame to throw a NullReferenceException.

**Comment by kingces95** [comment]
Could you please provide a zipped reproduction?

**Comment by tmt242001** [comment]
@kingces95 zipped reproduction
[CarouselViewIssue.zip](https://github.com/xamarin/Xamarin.Forms/files/3011068/CarouselViewIssue.zip)

**Comment by PureWeen** [comment]
The thing to keep in mind about visual is that it uses different renderers, controls, themes, and does things like wrap the Context class so there's a chance 3rd party libraries might not be compatible.

I ran the sample and the exception happens inside the CarouselViewRenderer 
```
03-26 23:10:24.529 I/MonoDroid( 7046):   at CarouselView.FormsPlugin.Android.CarouselViewRenderer.OnElementChanged (Xamarin.Forms.Platform.Android.ElementChangedEventArgs`1[TElement] e) [0x00132] in <892bb6a09e5d43ddbe9591d58bbf4c0e>:0 
03-26 23:10:24.529 I/MonoDroid( 7046):   at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].SetElement (TElement element) [0x000f5] in H:\github\Xamarin.Forms\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:228 
03-26 23:10:24.529 I/MonoDroid( 7046):   at Xamarin.Forms.Platform.Android.VisualElementRenderer`1[TElement].Xamarin.Forms.Platform.Android.IVisualElementRenderer.SetElement (Xamarin.Forms.VisualElement element) [0x00030] in H:\github\Xamarin.Forms\Xamarin.Forms.Platform.Android\VisualElementRenderer.cs:127 
```

Pinging @alexrainman in case he has thoughts on what inside that method could be causing the NRE

**Comment by lewcianci** [comment]
I don't know how to write this without sounding like I'm ungrateful, but it blows my mind that Visual could introduce those kinds of bugs to existing third party libraries.  I understand that every change in the Xamarin Forms codebase is one that could cause more issues further up the chain, or with third party libraries like (in this case) CarouselView.  But with CarouselView seeing sizeable deployment as a popular third party library, I would think that some testing would take place with popular third party controls to ensure that compatibility still remains.

I would also say that, if it is _possible_ that _some controls_ could be incompatible with what Visual does to the parent controls, then is Visual really fit for purpose in its current implementation?  In development terms, if you are going to put a spin on the existing Xamarin controls (like Frame, etc) then whatever you _do_ to these controls need to be 100% compatible with the base control.  If it's not, then you don't have a Frame, you have a MaterialFrame (or whatever you want to call it).  I know you can't test against every single third party library, but when something in XF is changed and it means that _another_ code maintainer has to go back and ensure that his code still works with the updates, that is a fundamental flaw.  Essentially it means that Visual isn't backwards compatible with any third party libraries, and you can't be sure that any third party library will work inside a Visual control.  What's *more* is that when this happens to you, you won't know its Visual causing this, meaning you have to spend hours working backwards until you decide to take the Visual tag out.  It also means that if I wanted to create my own custom Visual I run the risk of breaking an indeterminate amount of third party controls also.  So obviously, theres very little incentive for me to do that, and very little incentive to view that as a perk for XF.

TL;DR, Visual Frames (Visual anything) should be 100% compatible with the original control (in this case, Frames).  No amount of visual sugar is worth breaking an unknown amount of third party controls.

I don't mean to come across as cross or bombastic in this posting, and I certainly appreciate the huge effort that the XF team does in maintaining XF.  I also could completely misunderstand Visual in which case I'm sorry.

**Comment by andreinitescu** [comment]
> but when something in XF is changed and it means that another code maintainer has to go back and ensure that his code still works with the updates, that is a fundamental flaw.

@lewcianci I know you are talking to XF owners, but I'd like to share my feedback also:

First, what's the problem with the maintainer having to go back and update some things in his library? Isn't it natural to do such things? As developers we often need to go back to our code, refactor some things, break it and fix it. I find it a natural process.

Second, I'm in favor of seeing Xamarin Forms breaking things, as long as breaking comes from making Xamarin Forms more **feature rich**, more **stable** and run **faster**.
I don't agree with idea of avoiding breaking at all costs, it's extremely unproductive. Along the years, not having many things breaking in a project of this size and importance is not a good sign at all. 

And "breaking" comes in different forms. Just as an example, ListView has a lot of issues when using it in more than the basic scenario. CollectionView is great addition, but we all use ListView in our apps, so when CollectionView will be released, we'll have to go back and refactor our code to CollectionView, won't we?

**Comment by lewcianci** [comment]
> First, what's the problem with the maintainer having to go back and update some things in his library? Isn't it natural to do such things? As developers we often need to go back to our code, refactor some things, break it and fix it. I find it a natural process.

Because sometimes, some libraries are a long time between updates, or some don't see updates at all.  CarouselView is a good example of this - it broke with the XF 3.5 update (totally understandable, there were a lot of new features with that and breaking changes).  But it hasn't been updated since January 2018.  So, when alexrainman updated the dependency, it started working again.

It's okay for me and you to "go back and update our libraries" because today we are both active Xamarin Forms developers.  But not everyone is guaranteed to always be active.  I would say that CarouselView is used quite a lot, and the fact that something like adding a Tag to a control can straight up break a popular third party control is very strange.  Plus, to make things worse, the fix isn't intuitive.  Is there a "fast-track" to fixing third party libraries, or a list of things that Visual has changed that a library maintainer should look for? Would their component now be incompatible for whatever reason?

I think it should be part of XF's test suite to test a few popular third party libraries with very basic implementations.  Because at the moment it seems like things are happening with no regard to the potential impact it could have on third party libraries.

I just think that:

1. It's really hard to know that Visual is breaking CarouselView and not intuitive to a developer to _know_ that when developing and trying to work out why they are getting a NullReferenceException
2. Visual doesn't have enough "red flags" around it that it is _not_ a drop-in tag that makes things look different, the implications of it are far greater than that
3. It's not easy to know how to make your third party library "Visual Compatible"
4. A Frame should always behave like a Frame, no matter what Key is on it.  If it doesn't, and if it breaks for _whatever reason_, then it is not the same, and should be called a VisualFrame.

**Comment by davidortinau** [comment]
Hi all, I very much appreciate the discussion here. Thank you. 

To clarify, Visual isn't breaking anything. CarouselView was written against (or at the time of) default renderers. The new Material platform renderers (made easy to consume using the new Visual API) are what need to be accounted for.

I agree, it is a concern that this NRE came up and I would like to better understand what about the new renderer has caused this. 

This is similar in some respects to FastRenderers for Android where the XF API for those controls remained the same, but the platform renderers were vastly different (and improved). We provide a flag to put the choice in your hands - something I would like to emphasize. 

There will be some adjustment as the ecosystem matures in response to the ripples of the Xamarin platform changes. The choice of when and how to adapt is in your hands, and we will adapt with you along the way. 

So let's figure out what is triggering this NRE, how and if we can avoid it, and then decide how to act.

**Comment by PureWeen** [comment]
>  But with CarouselView seeing sizeable deployment as a popular third party library, I would think that some testing would take place with popular third party controls to ensure that compatibility still remains.

This is a good point and we do have work currently scheduled (https://github.com/xamarin/Xamarin.Forms/issues/5656) to run automated tests against 3rd party libraries (Telerik, Sync Fusion, FFImageLoading). CarouselView would be a good one to add to this. We do want to know when an upgrade breaks a 3rd party library and to quickly identify which side of the fence the issue comes from. Did we break an API? Are they making an incorrect assumption? How can we start coordination early on so everything is smooth.  We are actively addressing this.

> Visual doesn't have enough "red flags" around it that it is not a drop-in tag that makes things look different, the implications of it are far greater than that

This is a good point and I could see us adding some additional documentation #5718  with the material docs about what differences with the components people should expect to see.

> So let's figure out what is triggering this NRE, how and if we can avoid it, and then decide how to act.

This is mainly where I'm curious. I looked at the source for CarouselView and unfortunately didn't see anything that stood out to me so was just pinging Alex in case something was obvious to him. If I don't hear from him I'll look into this later this week to see if I can find the root cause. 

> I just want to reiterate that I really like XF and appreicate everyones hard work on it. I'm not trying to razz anyone up and certainly, Visual is a welcome addition.

Your comments are all very appreciated!!! Thank you for taking the time to make them.

**Comment by PureWeen** [comment]
@alexrainman
So the issue is here

https://github.com/alexrainman/CarouselView/blob/master/CarouselView/CarouselView.FormsPlugin.Android/CarouselViewImplementation.cs#L70

Context isn't going to be of type Activity
There's a null check in the CarouselView code for this but then the keyboard service is used later on in the code and causes an NRE

I've added an extension method in the latest forms here
https://github.com/xamarin/Xamarin.Forms/blob/3.6.0/Xamarin.Forms.Platform.Android/ContextExtensions.cs#L86

that can be used to retrieve the activity though this probably won't be super helpful directly since CarouselView is compiled against 2.5 but it can at least be used for reference

Also if possible the DebugType on the CarouselView projects should probably be set to *portable*


On a side note if you want this to currently work with CarouselView then just set the internal content of the CarouselView to Material instead of setting it on the main page
